File: benchmark/algos_bench/bench_algos.py
import sys
sys.setrecursionlimit(2_000_000)

def py_loop_100m():
    sum_val = 0
    for i in range(100_000_000):
        sum_val = (sum_val + i) ^ (i << 1)
    return sum_val

def py_fib(n):
    if n <= 1:
        return n
    return py_fib(n - 1) + py_fib(n - 2)

def py_prime_sieve():
    N = 10_000_000
    is_prime = [True] * (N + 1)
    is_prime[0] = is_prime[1] = False
    p = 2
    while p * p <= N:
        if is_prime[p]:
            for i in range(p * p, N + 1, p):
                is_prime[i] = False
        p += 1
    return sum(is_prime)

def py_quicksort_rec(arr, low, high):
    if low < high:
        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            if arr[j] < pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        pi = i + 1
        py_quicksort_rec(arr, low, pi - 1)
        py_quicksort_rec(arr, pi + 1, high)

def py_quicksort():
    N = 50_000
    seed = 12345
    arr = []
    for _ in range(N):
        seed = (seed * 1103515245 + 12345) & 0x7fffffff;
        arr.append(seed)
    py_quicksort_rec(arr, 0, N - 1)
    return arr[-1]

def main():
    mode = sys.argv[1] if len(sys.argv) > 1 else "loop"
    if mode == "py_loop":
        py_loop_100m()
    elif mode == "py_fib":
        py_fib(35)
    elif mode == "py_sieve":
        py_prime_sieve()
    elif mode == "py_sort":
        py_quicksort()

if __name__ == "__main__":
    main()

File: benchmark/algos_bench/bench_algos.js
function js_loop_100m() {
    let sum_val = 0;
    for (let i = 0; i < 100000000; ++i) {
        sum_val = (sum_val + i) ^ (i << 1);
    }
    return sum_val;
}

function js_fib(n) {
    if (n <= 1) return n;
    return js_fib(n - 1) + js_fib(n - 2);
}

function js_prime_sieve() {
    const N = 10000000;
    const is_prime = new Uint8Array(N + 1);
    is_prime.fill(1);
    is_prime[0] = is_prime[1] = 0;
    for (let p = 2; p * p <= N; ++p) {
        if (is_prime[p]) {
            for (let i = p * p; i <= N; i += p) {
                is_prime[i] = 0;
            }
        }
    }
    let count = 0;
    for (let i = 0; i <= N; ++i) {
        if (is_prime[i]) count++;
    }
    return count;
}

function js_quicksort_rec(arr, low, high) {
    if (low < high) {
        const pivot = arr[high];
        let i = low - 1;
        for (let j = low; j < high; ++j) {
            if (arr[j] < pivot) {
                i++;
                const t = arr[i]; arr[i] = arr[j]; arr[j] = t;
            }
        }
        const t = arr[i + 1]; arr[i + 1] = arr[high]; arr[high] = t;
        const pi = i + 1;
        js_quicksort_rec(arr, low, pi - 1);
        js_quicksort_rec(arr, pi + 1, high);
    }
}

function js_quicksort() {
    const N = 50000;
    const arr = new Int32Array(N);
    let seed = 12345;
    for (let i = 0; i < N; ++i) {
        seed = (seed * 1103515245 + 12345) & 0x7fffffff;
        arr[i] = seed;
    }
    js_quicksort_rec(arr, 0, N - 1);
    return arr[N - 1];
}

const mode = process.argv[2] || "js_loop";
if (mode === "js_loop") js_loop_100m();
else if (mode === "js_fib") js_fib(35);
else if (mode === "js_sieve") js_prime_sieve();
else if (mode === "js_sort") js_quicksort();
